### PR TITLE
Add skipmap flag to skip validation for AdditionlPropperties

### PR DESCRIPTION
**Pull request by Liujingfang1**
#137 

For `types.go` with content
```
package v1beta1

import (
    metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
)

type MyKindSpec struct {
    Replica int `json:"replica,omitempty"`
    Request map[string]string `json:"request,omitempty"`
}

type MyKindStatus struct {
    value int `json:"value,omitempty"`
}

// +genclient
// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object

// MyKind
// +k8s:openapi-gen=true
// +resource:path=mykinds
type MyKind struct {
    metav1.TypeMeta   `json:",inline"`
    metav1.ObjectMeta `json:"metadata,omitempty"`

    Spec   MyKindSpec   `json:"spec,omitempty"`
    Status MyKindStatus `json:"status,omitempty"`
}
```
The `install.yaml` file is
```
apiVersion: apiextensions.k8s.io/v1beta1
kind: CustomResourceDefinition
metadata:
  creationTimestamp: null
  labels:
    api: ""
    kubebuilder.k8s.io: master
  name: mykinds.jingfang.example.com
spec:
  group: jingfang.example.com
  names:
    kind: MyKind
    plural: mykinds
  scope: Namespaced
  validation:
    openAPIV3Schema:
      properties:
        apiVersion:
          type: string
        kind:
          type: string
        metadata:
          type: object
        spec:
          properties:
            replica:
              type: int
            request:
              type: object
          type: object
        status:
          properties:
            value:
              type: int
          type: object
      type: object
  version: v1beta1
status:
  acceptedNames:
    kind: ""
    plural: ""
  conditions: null
```
`kubectl apply -f install.yaml` successfully creates the CRD

### CODE REVIEW

File: cmd/kubebuilder/create/config/config.go [hunk 1]
```diff
@@ -75,4 +75,5 @@ func AddCreateConfig(cmd *cobra.Command) {
 	configCmd.Flags().StringVar(&controllerImage, "controller-image", "", "name of the controller container to run.")
 	configCmd.Flags().StringVar(&name, "name", "", "name of the installation.  used to generate the namespace and resource names.")
 	configCmd.Flags().StringVar(&output, "output", filepath.Join("hack", "install.yaml"), "location to write yaml to")
+	configCmd.Flags().BoolVar(&skipmap, "skip-map", true, "if set to true, skip validation of map in CRD")
```
Comment (droot) [inline cmd/kubebuilder/create/config/config.go]: May be rename the flag to "skip-map-validation"

Also, text can be more explicit:
`if set to true, skip generating validation schema for map type` ?

File: cmd/kubebuilder/create/config/gen.go [hunk 2]
```diff
@@ -36,7 +36,9 @@ import (
 )
 
 // CodeGenerator generates code for Kubernetes resources and controllers
-type CodeGenerator struct{}
+type CodeGenerator struct{
+	SkipMap bool
```
Comment (droot) [inline cmd/kubebuilder/create/config/gen.go]: SkipMap --> SkipMapValidation ? and in other places.

File: cmd/internal/codegen/parse/crd.go [hunk 3]
```diff
@@ -215,28 +215,36 @@ func (b *APIs) parsePrimitiveValidation(t *types.Type, found sets.String, commen
...
-			Schema: &additionalProps},
+			Schema: &additionalProps}
 	}
 
 	buff := &bytes.Buffer{}
-	if err := mapTemplate.Execute(buff, ""); err != nil {
+	if err := mapTemplate.Execute(buff, mapTempateArgs{result, parseOption.SkipMap}); err != nil {
 		log.Fatalf("%v", err)
 	}
 	return props, buff.String()
```
Comment (droot) [inline cmd/internal/codegen/parse/crd.go]: Can we write test for this or `create config` command ?

Comment (Liujingfang1) [reply cmd/internal/codegen/parse/crd.go]: Yes, we can. I will update the test.

File: cmd/kubebuilder/create/config/config.go (hunk 1 above)
Comment (Liujingfang1) [reply cmd/kubebuilder/create/config/config.go]: Sounds good. I will revise the description.

File: cmd/internal/codegen/parse/crd.go (hunk 3 above)
Comment (Liujingfang1) [reply cmd/internal/codegen/parse/crd.go]: Done

File: cmd/kubebuilder/create/config/config.go (hunk 1 above)
Comment (Liujingfang1) [reply cmd/kubebuilder/create/config/config.go]: Done

**Comment by Liujingfang1** [comment]
PTAL. I will add more tests for this flag once we have the testing setup for the existing projects as proposed in #181

File: cmd/internal/codegen/parse/crd.go [hunk 4]
```diff
@@ -215,28 +215,36 @@ func (b *APIs) parsePrimitiveValidation(t *types.Type, found sets.String, commen
...
+type mapTempateArgs struct {
+	Result string
+	SkipMap bool
+}
+
 var mapTemplate = template.Must(template.New("map-template").Parse(
 	`v1beta1.JSONSchemaProps{
     Type:                 "object",
-    AdditionalProperties: &v1beta1.JSONSchemaPropsOrBool{
+    {{if not .SkipMap}}AdditionalProperties: &v1beta1.JSONSchemaPropsOrBool{
```
Comment (droot) [inline cmd/internal/codegen/parse/crd.go]: lets change it to SkipMapValidation to make it consistent.

File: cmd/internal/codegen/parse/crd.go [hunk 5]
```diff
@@ -215,28 +215,36 @@ func (b *APIs) parsePrimitiveValidation(t *types.Type, found sets.String, commen
...
+	if !parseOption.SkipMapValidation {
+		props.AdditionalProperties = &v1beta1.JSONSchemaPropsOrBool{
 			Allows: true,
-			Schema: &additionalProps},
+			Schema: &additionalProps}
 	}
 
 	buff := &bytes.Buffer{}
-	if err := mapTemplate.Execute(buff, ""); err != nil {
+	if err := mapTemplate.Execute(buff, mapTempateArgs{result, parseOption.SkipMapValidation}); err != nil {
```
Comment (droot) [inline cmd/internal/codegen/parse/crd.go]: Pl. use tagged syntax here.

File: cmd/kubebuilder-gen/codegen/run/generator.go [hunk 6]
```diff
@@ -45,12 +46,16 @@ func (g *CodeGenerator) AddResourceGenerator(generator codegen.ResourceGenerator
...
-type customArgs struct{}
-
 // Execute parses packages and executes the code generators against the resource and controller packages
 func (g *CodeGenerator) Execute() error {
 	arguments := args.Default()
-	arguments.CustomArgs = &customArgs{}
+
+	// Custom args.
+	customArgs := &parse.ParseOptions{}
+	pflag.CommandLine.BoolVar(&customArgs.SkipMapValidation, "skip-map", true,
```
Comment (droot) [inline cmd/kubebuilder-gen/codegen/run/generator.go]: skip-map --> skip-map-validation.

File: cmd/kubebuilder-gen/codegen/run/generator.go [hunk 7]
```diff
@@ -45,12 +46,16 @@ func (g *CodeGenerator) AddResourceGenerator(generator codegen.ResourceGenerator
...
+		"If set true, skip map types.")
```
Comment (droot) [inline cmd/kubebuilder-gen/codegen/run/generator.go]: skip map --> skip map validation

File: cmd/internal/codegen/parse/crd.go (hunk 4 above)
Comment (Liujingfang1) [reply cmd/internal/codegen/parse/crd.go]: Done

File: cmd/internal/codegen/parse/crd.go (hunk 5 above)
Comment (Liujingfang1) [reply cmd/internal/codegen/parse/crd.go]: Done

File: cmd/kubebuilder-gen/codegen/run/generator.go (hunk 6 above)
Comment (Liujingfang1) [reply cmd/kubebuilder-gen/codegen/run/generator.go]: Done

**Pull request merged**